### DISCUSSION TITLE
Information request about OpenSSL

### DISCUSSION THREAD

**Discussion by ZSHUser01** [openssl/openssl#24407] [Q&A] [answered]
Hi, 

I'm writing to ask a couple of question about the openssl library because I'm interestred in understanding how the TLS works. Specifically, I'm interestred to understand better if there's a place in the code where the code for generating the premaster key, master key and session key for the client side is located because I can't seem to find it. 

Thank you in advance for the patience and support

**Comment by mattcaswell** [comment, marked as answer]
This varies a bit by protocol. You're probably mainly interested in the `ssl_generate_master_secret`, `ssl3_generate_master_secret`, `tls1_generate_master_secret` and `tls13_generate_master_secret` functions

**Reply by ZSHUser01** [reply]
Hi Matt, thank you for providing me the information. I will check this functions that you've mentioned in the source code. 

Thank you again for the patience and support
